Perfetto tracing

samchesney opened the pull request:
- **fix(steam-perfetto-sys): check for non-zero exits in build script**
- **fix(steam-perfetto-sys): use non-standard "compatibility" option**
- **fix(steam-perfetto-sys): Avoid use of non-standard `-F` option**
- **fix(steam-perfetto-sys): improve robustness of build triggers**
- **feat(steam-perfetto): add build script and library**
- **refactor(steam-track): disambiguate schema references**
- **feat(steam-track,sim-ring): add Perfetto tracker MVP**
- **refactor(steam-spotter): use PathBufs rather than Strings for paths**
- **feat(steam-spotter): --perfetto option to generate MVP Perfetto trace**
- **chore: temporarily disable semver checks for steam-perfetto**

samchesney committed:
feat(steam-track,sim-ring): add Perfetto tracker MVP

Support for creating the required Perfetto Protobuf messages is provided
via the new perfetto_trace_builder module within steam-track. This
module enables code reuse with the tracker and future use cases such as
being able to post-process steam-track data into Perfetto traces.

The Perfetto tracker is considered to be the MVP, as whilst it currently
creates traces well suited for understanding data such as buffer
fullness from a simulation, it only creates Perfetto TrackEvents for
steam-track's Track enter and exit events.

The sim-ring example application has been extended to support the use of
the Perfetto tracker, allowing it to generate Perfetto traces, as well
as the previously supported text and binary trace formats supported by
steam-spotter.

This new functionality of steam-track is guarded by the `perfetto`
feature. With this new feature enabled the build of steam-track will
cause the steam-perfetto-sys crate build to download Perfetto from
source, and the steam-perfetto crate to generate the required Protobuf
accessor functions.

As the `_perfetto-protoc` feature of steam-perfetto is not enabled by
steam-track a Protobuf compiler must already available for use on the
path.

samchesney committed:
refactor(steam-track): disambiguate schema references

With the future use of Protobuf schemas possible, e.g. for Perfetto
support, this makes it clear the existing functions are to provide
Cap'n Proto schema support.

samchesney committed:
refactor(steam-spotter): use PathBufs rather than Strings for paths

samchesney committed:
feat(steam-spotter): --perfetto option to generate MVP Perfetto trace

steam-spotter can now generate Perfetto trace output which can be
visualised using the web-based Perfetto UI. When generating a Perfetto
trace the usual TUI will not be started.

Support for creating the required Perfetto Protobuf messages is provided
via the perfetto_trace_builder module within steam-track.

The Perfetto generator is considered to be the MVP, as whilst it
currently creates traces well suited for understanding data such as
buffer fullness from a simulation, it only creates Perfetto TrackEvents
for steam-track's Track enter and exit events.

This new functionality of steam-track is guarded by the `perfetto`
feature, which is enabled by default. With this new feature enabled the
build of steam-track will cause the steam-perfetto-sys crate build to
download Perfetto from source, and the steam-perfetto crate to generate
the required Protobuf accessor functions.

As the `_perfetto-protoc` feature of steam-perfetto is not enabled by
steam-track a Protobuf compiler must already available for use on the
path.

samchesney committed:
feat(steam-perfetto): add build script and library

steam-perfetto generates and exposes Perfetto Protobuf access functions.

The build script uses prost-build to generate Rust access functions from
the trace schema included in the Perfetto source, provided by the
steam-perfetto-sys package.

The Perfetto build system will build the Protobuf compiler from source
if used (e.g. when the `_perfetto_ui` feature is enabled).
steam-perfetto will use the Protobuf compiler provided by Perfetto (and
ensure it is built) when the `_perfetto-protoc` feature is enabled.

pthedinger reviewed (approved):
Generally looks really good. Well done.

samchesney committed:
fix(steam-perfetto-sys): check for non-zero exits in build script

samchesney committed:
fix(steam-perfetto-sys): use non-standard "compatibility" option

Switch from using `-h` to `-n` option. The `-n` option is provided for
compatibility with other ln implementations, such as the the one in the
GNU coreutils.

Previously the symlink creation would work as expected on macOS but fail
on Linux.

samchesney committed:
fix(steam-perfetto-sys): Avoid use of non-standard `-F` option

The FreeBSD (as used by macOS) and GNU implementations of `ln` both
support the `-F` option but unfortunately use it for completely
different behaviours.

When using the FreeBSD implementation `-F` implies `-f` (when `-i` is
not specified), and results the behaviour required here. The behaviour
of the `-f` option is common to both implementations, so is now set
explicitly instead.

samchesney committed:
fix(steam-perfetto-sys): improve robustness of build triggers

The robustness of various of the build triggers which rely on the
symlink created by the build script has been improved.

A trigger based on the symlink will be now also added when just the
_perfetto_src feature is enabled.

Explicit rebuild trigger on build.rs removed as it is handled
automatically by Cargo.

samchesney committed:
chore: temporarily disable semver checks for steam-perfetto

The semantic version checks are performed using the main branch as the
base version as the packages are not currently published to a registry.

Unfortunately cargo-semver-checks cannot not handle the case of a new
package being added on a branch and therefore not being found on main.
To work around this limitation we must disable the checks for new
packages when they are first added.

This commit should be reverted once steam-perfetto has been merged onto
the main branch.
